cowbird/monitoring/monitoring.py

```diff
@@ -1,177 +1,103 @@
+    def __init__(self, config=None):
+        # type: (AnySettingsContainer) -> None
+        """
+        Initialize the monitoring instance from configured application settings.
+
+        @param config: AnySettingsContainer object from which the db can be retrieved.
+                       The default value of None is only there to disable pylint E1120. The singleton instance
+                       must be initialized with a proper config and after that the init function should not be hit.
+        """
+        if not config:
+            raise Exception("Without proper application settings, the Monitoring class cannot obtains a proper database"
+                            " store.")
         self.monitors = defaultdict(lambda: {})
+        self.store = get_db(config).get_store(MonitoringStore)
 
     def start(self):
         """
-        .. todo:: Load and start monitors from the BD
+        Load existing monitors and start the monitoring.
         """
+        monitors = self.store.list_monitors()
+        for mon in monitors:
+            self.monitors[mon.path][mon.callback] = mon
+            mon.start()
 
     def register(self, path, recursive, cb_monitor):
-        # type: (str, bool, FSMonitor) -> None
+        # type: (str, bool, Union[FSMonitor, Type[FSMonitor], str]) -> Monitor
         """
         Register a monitor for a specific path and start it.
 
         @param path: Path to monitor
         @param recursive: Monitor subdirectory recursively?
-        @param cb_monitor: Events are sent to this FSMonitor object
+        @param cb_monitor: FSMonitor for which an instance is created and events are sent
+                           Can be an object, a class type implementing FSMonitor or a string containing module and class
+                           name.
         """
         try:
-            self.monitors[path][cb_monitor]
+            callback = Monitor.get_qualified_class_name(Monitor.get_fsmonitor_instance(cb_monitor))
+            monitor = self.monitors[path][callback]
+            # If the monitor already exists but is not recursive, make it recursive if required
+            # (recursivity takes precedence)
+            if not monitor.recursive and recursive:
+                monitor.recursive = True
+            return monitor
         except KeyError:
             # Doesn't already exist
-            mon = Monitor(path, recursive, cb_monitor)
-            self.monitors[path][cb_monitor] = mon
-            mon.save()
-            mon.start()
+            try:
+                mon = Monitor(path, recursive, cb_monitor)
+                mon.start()
+                self.monitors[mon.path][mon.callback] = mon
+                self.store.save_monitor(mon)
+                return mon
+            except MonitorException as exc:
+                LOGGER.warning("Failed to start monitoring the following path [%s] with this monitor [%s] : [%s]",
+                               path, callback, exc)
```

Comment:
If monitor save fails, is it still valid to return it although it started properly?
Is there a possibility that a following (and conflicting) monitor definition overrides the previous, but doesn't ever get stored since `self.monitors[path][callback]` will contain a value?

Reply:
> If monitor save fails, is it still valid to return it although it started properly?

We could return it, but when the service will restart it will be lost. Is it better to get the misbehavior asap or at reboot?
Right now if the save_monitor raise something it will interupt the event handling for the current service and perform an exception callback to Magpie if relevant.

> Is there a possibility that a following (and conflicting) monitor definition overrides the previous, but doesn't ever get stored since self.monitors[path][callback] will contain a value?

You're right! if a recursive callback is registered over a non-recursive one it was not saved. I'll fix that.